sdk/nodejs/runtime/closure/createClosure.ts

```diff
@@ -840,114 +871,94 @@ function getOrCreateEntry(
-            // after we've walked as much of the graph synchronously as possible.
-            context.asyncWorkQueue.push(async () => {
-                const val = await obj.promise();
-                const oldFrames = context.frames;
-                context.frames = framesCopy;
-                entry.output = getOrCreateEntry(new SerializedOutput(val), undefined, context, serialize, logInfo);
-                context.frames = oldFrames;
-            });
+        else if (await isOutputAsync(obj)) {
+            entry.output = await createOutputEntryAsync(obj);
         }
         else if (obj instanceof Promise) {
-            // captures the frames up to this point. so we can give a good message if we
-            // fail when we resume serializing this promise.
-            const framesCopy = context.frames.slice();
-
-            // Push handling this promise to the async work queue.  It will be processed in batches
-            // after we've walked as much of the graph synchronously as possible.
-            context.asyncWorkQueue.push(async () => {
-                const val = await obj;
-                const oldFrames = context.frames;
-                context.frames = framesCopy;
-                entry.promise = getOrCreateEntry(val, undefined, context, serialize, logInfo);
-                context.frames = oldFrames;
-            });
+            const val = await obj;
+            entry.promise = await getOrCreateEntryAsync(val, undefined, context, serialize, logInfo);
         }
         else if (obj instanceof Array) {
-            // Recursively serialize elements of an array. Note: we use getOwnPropertyNames as the array
-            // may be sparse and we want to properly respect that when serializing.
+            // Recursively serialize elements of an array. Note: we use getOwnPropertyNames as the
+            // array may be sparse and we want to properly respect that when serializing.
             entry.array = [];
-            for (const key of Object.getOwnPropertyNames(obj)) {
-                if (key !== "length" && obj.hasOwnProperty(key)) {
-                    entry.array[<any>key] = getOrCreateEntry(
-                        obj[<any>key], undefined, context, serialize, logInfo);
+            for (const descriptor of await getOwnPropertyDescriptors(obj)) {
+                if (descriptor.name !== undefined &&
+                    descriptor.name !== "length") {
+
+                    entry.array[<any>descriptor.name] = await getOrCreateEntryAsync(
+                        await getOwnPropertyAsync(obj, descriptor), undefined, context, serialize, logInfo);
                 }
             }
+
+            // TODO(cyrusn): It feels weird that we're not examining any other descriptors of an
+            // array.  For example, if someone put on a property with a symbolic name, we'd lose
+            // that here. Unlikely, but something we may need to handle in the future.
```

Comment:
You can totally do this in ES6:

```
class MyArray extends Array {
    constructor(...args) {
        super(...args);
        this.foo = 42;
    }
}
```

It's not that strange of a thing to do in JS, I think. can we log a bug for this?

Reply:
Yes.  I'll log an issue.  Though note that your case will work *fine*.  'foo' will be found as a descriptor on the array..

The cases that don't work properly are specifically:
1. you have a different *prototype* than 'array'.  i.e. you derived from a subclass of array.
2. you had a symbolicly named property attached to you.  

I will open bug to track both of these.

Reply:
https://github.com/pulumi/pulumi/issues/2139